### DISCUSSION TITLE
Unable to Make the TLS work, using PAid SSL and Have Docker for OvenMedia

### DISCUSSION THREAD

**Discussion by shoaibzafar** [OvenMediaLabs/OvenMediaEngine#1324] [Help] [answered]
Hi I am to make TLS configuration working and getting this error.

E [OvenMediaEngine:1] Certificate | certificate.cpp:47 | Failed to create a certificate for VirtualHost [default]:
Reason: [default] Could not create a certificate from file - [OpenSSL] error:80000002:system library::No such file or directory (-2147483646)

     Cert file path: origin_conf/osng01.multiplierapp.live.crt
    Chain cert file path: origin_conf/osng01.multiplierapp.live.ca-bundle
    Private key file path: origin_conf/privkey.pem
    
    FYI I have please all the files with 777 permission at location.
    
    <Host>
			<Names>
				
				<Name>osng01.multiplierapp.live</Name>
			</Names>
			<TLS>
				<CertPath>osng01.multiplierapp.live.crt</CertPath>
				<KeyPath>privkey.pem</KeyPath>
				<ChainCertPath>osng01.multiplierapp.live.ca-bundle</ChainCertPath>
			</TLS>
		</Host>

**Comment by dimiden** [comment, marked as answer]
@shoaibzafar
For accurate root cause analysis, please check the following:

1. It seems like you are running OME in Docker. Please verify that the certificate files are mounted correctly.
2. If the issue persists despite the certificate files being correctly mounted, please upload the full logs of OME. By examining the full logs, I can identify the path to your OME's ```Server.xml``` file, etc.

**Reply by shoaibzafar** [reply]
Thank you @dimiden  I got it working. I really Appreciate the community is really helpful. However I am facing another issue that whenever I try to enable the configuration for API Server OME seems to stopped working. Please let me know if you need any information to help me with API Server working.

**Reply by dimiden** [reply]
@shoaibzafar 
There can be various reasons why the API server may not be functioning, such as firewall settings, lack of port forwarding, or incorrect configurations.
Therefore, based on the information you provided, we cannot determine the exact cause of the API server not working.

**Reply by shoaibzafar** [reply]
Hi @dimiden  There is no firewall working it is test server so kept all ports open.

About the configuration let me share exact tags with you. 

<Bind>
		
		
		<Managers>
			<API>
				<Port>${env:OME_API_PORT:8081}</Port>
				<WorkerCount>1</WorkerCount>
			</API>
		</Managers>

**Reply by dimiden** [reply]
@shoaibzafar 
How did you call that API? Does OME generate logs when the API is called? What options did you specify when running the Docker container?
As I mentioned earlier, since there can be various causes, I would appreciate it if you could provide as much information as possible.

**Reply by shoaibzafar** [reply]
I am executing the Docker as ./ome_docker_launcher.sh restart after making changes in Server.xml I followed this link to deploy the Oven Media  https://airensoft.gitbook.io/ovenmediaengine/getting-started  
Beside this I am attaching the server.xml file and logs file below. Please check that. 
[ovenmediaengine.log]

(https://github.com/AirenSoft/OvenMediaEngine/files/12102469/ovenmediaengine.log)
[Server.xml.txt](https://github.com/AirenSoft/OvenMediaEngine/files/12102480/Server.xml.txt)

**Reply by getroot** [reply]
Your Managers Bind configure is commented out
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/17779085/8c22be65-0ca7-44bb-8b00-0f3eb681a589)

And there is no `<Managers>` config like this https://airensoft.gitbook.io/ovenmediaengine/rest-api#managers

Please see the manual for detail

**Reply by shoaibzafar** [reply]
When I Uncomment this above mentioned tags, nothing works. Can you please share the exact configuration keep in mind my server.xml files. thank you a lot !

**Reply by getroot** [reply]
Did you add `<Managers>` configuration (NOT `<Bind><Managers>`) in your Server.xml? Please see the manual link I provided 
 https://airensoft.gitbook.io/ovenmediaengine/rest-api#managers

**Reply by shoaibzafar** [reply]
I got no error now as you directed the configuration. Thank you very much for the support. Just one last thing how can I access the api now. 
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/11497608/4938a317-cfe6-43c0-83f7-86082b3dc790)
Access the api like shown in screenshot above

**Reply by dimiden** [reply]
OME doesn't respond an HTML. Please check your network environment.

**Reply by getroot** [reply]
That response doesn't look like from OvenMediaEngine but Numble

![image](https://github.com/AirenSoft/OvenMediaEngine/assets/17779085/ef3baa7a-16b9-4ff7-af93-9a8f01612f3e)

**Reply by shoaibzafar** [reply]
Thank you @getroot  for pointing out that, I disabled the nimble Streamer. 
Now I am facing the below issue. 
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/11497608/a42afc17-41ed-41f8-9707-9cb340d7ba87)

Here you can see the API server is up and running. 
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/11497608/bd4a3398-0c8a-446a-8f3e-e5e63000c28d)

**Reply by getroot** [reply]
Try version 0.15.14.

https://github.com/AirenSoft/OvenMediaEngine/issues/1306

**Reply by getroot** [reply]
To be precise, the file below should have been updated. Download again and try.

https://raw.githubusercontent.com/AirenSoft/OvenMediaEngine/master/misc/ome_docker_launcher.sh

**Reply by shoaibzafar** [reply]
do I need to do the setup again or just restart that?

**Reply by dimiden** [reply]
@shoaibzafar 
You don't need to ```setup``` again, just ```stop``` and then ```start``` using the script. (Note - not ```restart```)

**Reply by shoaibzafar** [reply]
@dimiden  @getroot  Thank you very both of you, I am able to get the things working. 
Much Appreciate your timely responses. Love the OME Community

**Reply by jfinkel** [reply]
"Please verify that the certificate files are mounted correctly."

Would you kindly tell me what this means?  I have valid certs that I got via Certbot.  They work fine in an nginx server.   I copied them into the running OME Docker container and put them in `/etc/ssl/private/` from which they are referenced in Server.xml:

```
<TLS>
    <CertPath>/etc/ssl/private/cert.pem</CertPath>
   <KeyPath>/etc/ssl/private/privkey.pem</KeyPath>
   <ChainCertPath>/etc/ssl/private/chain.pem</ChainCertPath>
</TLS>
```
Are you saying that they must stay within the Docker host's filesystem and mounted to bind with an internal directory via docker's -v directive?

By the way, the logs indicate that the certs were successfully loaded:

```
[2024-01-04 20:45:15.278] I [OvenMediaEngine:1] Certificate | certificate.cpp:35   | A certificate has been loaded for VirtualHost [default]:
	Cert file path: /etc/ssl/private/cert.pem
	Chain cert file path: /etc/ssl/private/chain.pem
	Private key file path: /etc/ssl/private/privkey.pem
```

But attempting to use either the wss:  or https: protocols to access the stream fail.

Thanks in advance.
